text/0000-bench-utils.md

```diff
@@ -0,0 +1,208 @@
+- Feature Name: black_box
+- Start Date: 2018-03-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC adds one function, `core::hint::black_box`, which is a hint to the
+optimizer to disable certain compiler optimizations. 
+
+# Motivation
+[motivation]: #motivation
+
+A tool for preventing compiler optimizations is widely useful. One application
+is writing synthetic benchmarks, where, due to the constrained nature of the
+benchmark, the compiler is able to perform optimizations that wouldn't otherwise
+trigger in practice. Another application is writing constant time code, where it
+is undesirable for the compiler to optimize certain operations depending on the
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can use `x` in any possible valid way that Rust code is
+allowed to without introducing undefined behavior in the calling code. This
+requires the compiler to be maximally pessimistic in terms of optimizations, but
+the compiler is still allowed to optimize the expression generating `x`. This
+function returns `x` and is a no-op in the virtual machine.
```

Comment:
I still disagree about even mentioning a virtual machine. I expect most people will think of [something like this](https://www.virtualbox.org/).

Instead, I'd add a new paragraph after this one, saying something like

> However, while the *compiler* must assume that `black_box` performs any legal mutation of `x`, the *programmer* can rely on `black_box` not actually having any effect (other than inhibiting optimizations).

Reply:
I've re-ordered this, removing the mention of the virtual machine, and incorporating your comment into it.